src/radiation/moments.cpp

```diff
@@ -846,54 +847,46 @@ TaskStatus CalculateGeometricSourceImpl(T *rc, T *rc_src) {
           Vec covH{{J * v(iblock, idx_H(ispec, 0), k, j, i),
                     J * v(iblock, idx_H(ispec, 1), k, j, i),
                     J * v(iblock, idx_H(ispec, 2), k, j, i)}};
-
-          Vec conF;
-          g.raise3Vector(covF, &conF);
-
-          Real con_H4[4] = {0};
-          SPACELOOP2(ii, jj) { con_H4[ii + 1] += g.con_gamma(ii, jj) * covH(jj); }
-
-          Real conTilPi[4][4] = {0};
+          Tens2 conTilPi;
           if (iTilPi.IsValid()) {
             SPACELOOP2(ii, jj) {
-              conTilPi[ii + 1][jj + 1] = v(iblock, iTilPi(ispec, ii, jj), k, j, i);
+              conTilPi(ii, jj) = v(iblock, iTilPi(ispec, ii, jj), k, j, i);
             }
           } else {
-            Tens2 con_tilPi{0};
-            c.GetConTilPiFromPrim(J, covH, &con_tilPi);
-            SPACELOOP2(ii, jj) { conTilPi[ii + 1][jj + 1] = con_tilPi(ii, jj); }
-          }
-
-          Real con_T[4][4];
-          SPACETIMELOOP2(mu, nu) {
-            Real conh_munu = con_g[mu][nu] + con_u[mu] * con_u[nu];
-            con_T[mu][nu] = (con_u[mu] * con_u[nu] + conh_munu / 3.) * J;
-            con_T[mu][nu] += con_u[mu] * con_H4[nu] + con_u[nu] * con_H4[mu];
-            con_T[mu][nu] += J * conTilPi[mu][nu];
+            c.GetConTilPiFromPrim(J, covH, &conTilPi);
           }
+          Tens2 conP;
+          c.getConPFromPrim(J, covH, conTilPi, &conP);
 
-          // TODO(BRR) add Pij contribution
+          Vec conF;
+          g.raise3Vector(covF, &conF);
 
-          Real srcE = 0.0;
+          Real con_T[4][4] = {0};
+          SPACETIMELOOP2(mu, nu) { con_T[mu][nu] += con_n[mu] * con_n[nu] * E; }
           SPACETIMELOOP(mu) {
-            srcE += con_T[mu][0] * dlnalp[mu];
-            SPACETIMELOOP(nu) {
-              Real Gamma_udd = 0.;
-              SPACETIMELOOP(lam) { Gamma_udd += con_g[0][lam] * Gamma[lam][nu][mu]; }
-              srcE -= con_T[mu][nu] * Gamma_udd;
+            SPACELOOP(ii) {
+              con_T[mu][ii + 1] += con_n[mu] * conF(ii);
+              con_T[ii + 1][mu] += con_n[mu] * conF(ii);
```

Comment:
The previous evaluation of Tmunu for the radiation was incorrect because it assumed H^mu = Htilde^mu

Reply:
Good catch, I agree it makes more sense to just work with the lab frame quantities.